fidouafclient/app/src/main/java/org/ebayopensource/fidouafclient/op/Reg.java

```diff
@@ -38,17 +38,22 @@
 public class Reg {
 	
 	private Gson gson = new GsonBuilder().disableHtmlEscaping().create();
-	
+
 	public String getUafMsgRegRequest (String username, String appId){
```

Comment:
Hi,
I had a look to the code (client and server), FIDO specifications and I think you are right about misuse of these terms (appID and facetID), but I'm not sure if MainActivity needs to be changed. Probably the candidates are [RegistrationRequestProcessor](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fido/uaf/client/RegistrationRequestProcessor.java#L35) and [AuthenticationRequestProcessor](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fido/uaf/client/AuthenticationRequestProcessor.java#L31).

Let me try to explain.

According to [FIDO Glossary](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-glossary-v1.0-ps-20141208.html):

- **Application Facet** is how an application is implemented on various platforms. For example, the application MyBank may have an Android app, an iOS app, and a Web app. These are all facets of the MyBank application.
- **Application Facet ID (facetID)** is a platform-specific identifier (URI) for an application facet. For Web applications, the facet id is the RFC6454 origin. For Android applications, the facet id is the URI android:apk-key-hash:<hash-of-apk-signing-cert>
- **AppID** is an identifier for a set of different Facets of a relying party's application. The AppID is a URL pointing to the TrustedFacets, i.e. list of FacetIDs related to this AppID.

According to [FIDO UAF Protocol](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-uaf-protocol-v1.0-ps-20141208.html)
 
In [section 3.4.1 "Registration Request Message"](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-uaf-protocol-v1.0-ps-20141208.html#registration-request-message) we have: 

```
[{"header": {"upv": {"major": 1,"minor": 0},
 "op": "Reg",
 "appID": "https://uaf-test-1.noknoktest.com:8443/SampleApp/uaf/facets",
 "serverData":.............
```

And after execution of [getRegRequest method in Reg class](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fidouafclient/op/Reg.java#L45) we have:

```
[{"header": {"upv": {"major": 1,"minor": 0},
 "op": "Reg",
 "appID": "http:\/\/www.head2toes.org\/fidouaf\/v1\/public\/uaf\/facets",
 "serverData":.............
```

So far so good. In [section 3.4.4 Registration Response Message](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-uaf-protocol-v1.0-ps-20141208.html#registration-request-message) we have:

```
[{"assertions": [{..............
"header": {
"appID": "https://uaf-test-1.noknoktest.com:8443/SampleApp/uaf/facets",
"op": "Reg",
```

However after execution of [clientSendRegResponse method in Reg class](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fidouafclient/op/Reg.java#L90) we have:


```
[{"assertions": [{..............
"header": {
"appID": "android:apk-key-hash:Lir5oIjf552K/XN4bTul0VS3GfM",
"op": "Reg",
```

So, you are right. There is something wrong here!

In [section 3.4.6.2 Registration Request Processing Rules for FIDO UAF Clients](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-uaf-protocol-v1.0-ps-20141208.html#registration-request-message)

The **FIDO UAF Client MUST** perform the following steps:

* Obtain **FacetID** of the requesting Application. If the **AppID** is missing or empty, set the **AppID** to the **FacetID**.
  * Verify that the **FacetID** is authorized for the **AppID** according to the algorithms in [FIDOAppIDAndFacets](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-appid-and-facets-v1.0-ps-20141208.html#the-appid-and-facetid-assertions).
  * If the **FacetID** of the requesting Application is not authorized, reject the operation

So, current [getUafMsgRegRequest](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fidouafclient/op/Reg.java#L47) is not doing that.

In my pull request I was trying (a really naive way) to "Verify that the FacetID is authorized for the AppID according to the algorithms in [FIDOAppIDAndFacets](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-appid-and-facets-v1.0-ps-20141208.html#the-appid-and-facetid-assertions).", assuming that appId param as acting as facetID.

Deepening a little more, In [section 3.4.6.4 Registration Response Generation Rules for FIDO UAF Client](https://fidoalliance.org/specs/fido-uaf-v1.0-ps-20141208/fido-uaf-protocol-v1.0-ps-20141208.html#registration-response-generation-rules-for-fido-uaf-client)

The **FIDO UAF Client MUST** follow the steps:

* Create a `RegistrationResponse` message
* Copy `RegistrationRequest.header` into `RegistrationResponse.header`

The **FIDO Server MUST** follow the steps: 

* Verify each field in `fcp` (FinalChallengeParams) and make sure it is valid: 
  * Make sure `fcp.appID` corresponds to the one stored by the FIDO Server
  * Make sure `fcp.facetID` is in the list of trusted FacetIDs

However in [processRequest method in RegistrationRequestProcessor class](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fido/uaf/client/RegistrationRequestProcessor.java#L51) we are sending these values to FIDO Server:

* appID = android:apk-key-hash:Lir5oIjf552K/XN4bTul0VS3GfM; and
* facetId = null;

So, I think [this line is not correct in fidouafclinet.op.Reg class](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fidouafclient/op/Reg.java#L47) because it is overwritten "appId" param in the request message's header, considering that **FIDO Client MUST** Copy `RegistrationRequest.header` into `RegistrationResponse.header`

Finally, I think that we need some work in `getFacetId` and `setAppId` methods in [RegistrationRequestProcessor class](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fido/uaf/client/RegistrationRequestProcessor.java) and in [AuthenticationRequestProcessor](https://github.com/eBay/UAF/blob/a9a14fe51eae6b4d8c319b965fb953b539aadcd1/fidouafclient/app/src/main/java/org/ebayopensource/fido/uaf/client/AuthenticationRequestProcessor.java), considering that **FIDO Server MUST** make sure `fcp.facetID` is in the list of trusted FacetIDs 

Well, probably I missed something :-). Please let me know what you think.